src/middleware.ts

```diff
@@ -216,44 +228,57 @@ export const persist = <S extends State>(
       blacklist.forEach((key) => delete state[key])
     }
 
-    storage?.setItem(name, await serialize({ state, version }))
+    return storage?.setItem(name, await serialize({ state, version }))
   }
 
   const savedSetState = api.setState
 
   api.setState = (state, replace) => {
     savedSetState(state, replace)
-    setItem()
+    return setItem()
   }
 
   // rehydrate initial state with existing stored state
   ;(async () => {
     const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
 
+    let deserializedStorageValue
     try {
-      const storageValue = await storage.getItem(name)
-
+      const storageValue = await storage?.getItem(name)
       if (storageValue) {
-        const deserializedStorageValue = await deserialize(storageValue)
+        deserializedStorageValue = await deserialize(storageValue)
+      }
+    } catch (e) {
+      onRehydrateError?.(e)
+    }
 
-        // if versions mismatch, clear storage by storing the new initial state
-        if (deserializedStorageValue.version !== version) {
-          setItem()
-        } else {
-          set(deserializedStorageValue.state)
+    if (deserializedStorageValue) {
+      // if versions mismatch, run migration
+      if (deserializedStorageValue.version !== version) {
+        try {
+          const migratedState = await migrate?.(
+            deserializedStorageValue.state,
+            deserializedStorageValue.version
+          )
+          if (migratedState) {
+            set(migratedState)
+            await setItem()
+          }
+        } catch (e) {
+          onRehydrateError?.(e)
```

Comment:
Is there a reason for using two try catch here?

Reply:
Initially they throw two different error object. One error object is instantiation from `RehydrateError` and the second is from `MigrationError`, both are custom class extends from `Error`. I did that to identify where the error occured. And then I decided to pass the original error object rather than instantiating new error object to retain stack trace. So, now there's no need for two try catch blocks. I will combine them together.